aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,358 @@
+        let i = 0;
+        let len = vector::length(bucket);
+        while (i < len) {
+            let entry = vector::borrow(bucket, i);
+            assert!(&entry.key != &key, error::invalid_argument(EALREADY_EXIST));
+            i = i + 1;
+        };
+        let e = Entry {hash, key, value};
+        if (table.bucket_size == 0) {
+            assert!(table.size == 0, error::internal(EINVARIABLE_BROKEN));
+            let estimated_entry_size = max(size_of_val(&e), 1);
+            table.bucket_size = max(1024 /* free_write_quota */ / estimated_entry_size, 1);
+        };
+        vector::push_back(bucket, e);
+        table.size = table.size + 1;
+
+        if (load_factor(table) >= (table.split_load_threshold as u64)) {
+            split_one_bucket(table);
+        }
+    }
+
+    /// Split the next bucket into two and re-insert existing items.
+    fun split_one_bucket<K, V>(table: &mut SmartTable<K, V>) {
+        let new_bucket_index = table.num_buckets;
+        // the next bucket to split is num_bucket without the most significant bit.
+        let to_split = new_bucket_index ^ (1 << table.level);
+        let new_bucket = vector::empty();
+        table.num_buckets = new_bucket_index + 1;
+        // if the whole level is splitted once, bump the level.
+        if (to_split + 1 == 1 << table.level) {
+            table.level = table.level + 1;
+        };
+        let old_bucket = table_with_length::borrow_mut(&mut table.buckets, to_split);
+        // partition the bucket. after the loop, i == j and [0..i) stays in old bucket, [j..len) goes to new bucket
+        let i = 0;
+        let j = vector::length(old_bucket);
+        let len = j;
+        while (i < j) {
+            let entry = vector::borrow(old_bucket, i);
+            let index = bucket_index(table.level, table.num_buckets, entry.hash);
+            if (index == new_bucket_index) {
+                j = j - 1;
+                vector::swap(old_bucket, i, j);
+            } else {
+                i = i + 1;
+            };
+        };
+        while (j < len) {
+            let entry = vector::pop_back(old_bucket);
+            vector::push_back(&mut new_bucket, entry);
+            len = len - 1;
+        };
+        table_with_length::add(&mut table.buckets, new_bucket_index, new_bucket);
+    }
+
+    /// Return the expected bucket index to find the hash.
+    fun bucket_index(level: u8, num_buckets: u64, hash: u64): u64 {
+        let index = hash % (1 << (level + 1));
+        if (index < num_buckets) {
+            // in existing bucket
+            index
+        } else {
+            // in unsplitted bucket
+            index % (1 << level)
+        }
+    }
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow<K: drop, V>(table: &SmartTable<K, V>, key: K): &V {
+        let index = bucket_index(table.level, table.num_buckets, sip_hash_from_value(&key));
```

Comment:
How does this work if we inserted an item into an existing bucket (based on sip hash) but then that bucket got split? Can we find that item if it could have been moved into the new (split) bucket)?

Reply:
Can you elaborate more? Not sure what you mean by "transparent to users"

Reply:
I mean we can find them after splitting because you will first compute a `index A` with base x and find A >= the next index of bucket to be split, then you will recompute a `index B` which choose a different base (x / 2) and find the bucket.